js/markdown-stream/markdown-stream.ts

```diff
@@ -0,0 +1,257 @@
+
+      hljs.highlightElement(el);
+
+      // Add copy button
+      const btn = createElement("button", {
+        class: "code-copy-button",
+        title: "Copy to clipboard",
+      });
+      btn.innerHTML = '<i class="bi"></i>';
+      el.prepend(btn);
+
+      // Setup clipboard
+      const clipboard = new ClipboardJS(btn, { target: () => el });
+      clipboard.on("success", (e) => {
+        btn.classList.add("code-copy-button-checked");
+        setTimeout(
+          () => btn.classList.remove("code-copy-button-checked"),
+          2000
+        );
+        e.clearSelection();
+      });
+    });
+  }
+
+  // ------- Scrolling logic -------
+
+  // Nearest scrollable parent element (if any)
+  #scrollableElement: HTMLElement | null = null;
+  // Whether content is currently being added to the element
+  #isContentBeingAdded = false;
+  // Whether the user has scrolled away from the bottom
+  #isUserScrolled = false;
+
+  #onScroll = (): void => {
+    if (!this.#isContentBeingAdded) {
+      this.#isUserScrolled = !this.#isNearBottom();
+    }
+  };
+
+  #isNearBottom(): boolean {
+    const el = this.#scrollableElement;
+    if (!el) return false;
+
+    return el.scrollHeight - (el.scrollTop + el.clientHeight) < 50;
+  }
+
+  #updateScrollableElement(): void {
+    const el = this.#findScrollableParent();
+
+    if (el !== this.#scrollableElement) {
+      this.#scrollableElement?.removeEventListener("scroll", this.#onScroll);
+      this.#scrollableElement = el;
+      this.#scrollableElement?.addEventListener("scroll", this.#onScroll);
+    }
+  }
+
+  #findScrollableParent(): HTMLElement | null {
+    if (!this.auto_scroll) return null;
+
+    // eslint-disable-next-line
+    let el: HTMLElement | null = this;
+    while (el) {
+      if (el.scrollHeight > el.clientHeight) return el;
+      el = el.parentElement;
+    }
+    return null;
+  }
+
+  #maybeScrollToBottom(): void {
+    const el = this.#scrollableElement;
+    if (!el || this.#isUserScrolled) return;
+
+    el.scroll({
+      top: el.scrollHeight - el.clientHeight,
+      behavior: this.streaming ? "instant" : "smooth",
+    });
+  }
+
+  #cleanup(): void {
+    this.#scrollableElement?.removeEventListener("scroll", this.#onScroll);
+    this.#scrollableElement = null;
+    this.#isUserScrolled = false;
+  }
+}
+
+// ------- Register custom elements and shiny bindings ---------
+
+customElements.get("shiny-markdown-stream") ||
+  customElements.define("shiny-markdown-stream", MarkdownElement);
+
+function handleMessage(message: ContentMessage | IsStreamingMessage): void {
+  const el = document.getElementById(message.id) as MarkdownElement;
+
+  if (!el) {
+    showShinyClientMessage({
+      status: "error",
+      message: `Unable to handle MarkdownStream() message since element with id
+      ${message.id} wasn't found. Do you need to call .ui() (Express) or need a
+      output_markdown_stream('${message.id}') in the UI (Core)?`,
```

Comment:
It's currently

```suggestion
      markdown_stream_ui('${message.id}') in the UI (Core)?`,
```

but I like `output_markdown_stream()` better (unless you've already considered and moved away from this for reasons).

Reply:
Done in [add7b2f](https://github.com/posit-dev/py-shiny/pull/1782/commits/add7b2f48526adc67b8340ef77c2e76e8d4ac825)